### PR TITLE
Adversary as `pl.LightningModule`

### PR DESCRIPTION
**Pull request by dxoigmn**
# What does this PR do?

Right now we treat adversaries as special things with their own loops and callbacks, when really we should just treat them like LightningModules. Doing so means that we can just use a Trainer to fit its parameters. This PR attempts to make that so.

~~As of [dcf7599](https://github.com/IntelLabs/MART/pull/103/commits/dcf759916ba4a6e30e9549b894521427c12473ac), there is a bug in adversarial training.~~

Depends upon #146 and #147.

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] `pytest`
- [x] `python -m mart experiment=CIFAR10_CNN_Adv trainer=gpu` achieves 71% accuracy.
- [x] `python -m mart experiment=CIFAR10_CNN_Adv trainer=ddp datamodule.world_size=2 trainer.devices=2` achieves 71% accuracy.

## Before submitting

- [x] The title is **self-explanatory** and the description **concisely** explains the PR
- [x] My **PR does only one thing**, instead of bundling different changes together
- [x] I list all the **breaking changes** introduced by this pull request
- [x] I have commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by dxoigmn** [comment]
I started splitting this PR into pieces. I think I'll also try to stick to the same config structure.

File: mart/attack/perturber.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+if TYPE_CHECKING:
+    from .initializer import Initializer
+    from .objective import Objective
+    from .threat_model import ThreatModel
+
+__all__ = ["LitPerturber"]
+
+
+class LitPerturber(pl.LightningModule):
```
Comment (dxoigmn) [inline mart/attack/perturber.py:25]: Why can't this be a `LitModular`? Then we inherit all of the optimizer functionality. The problem, I think, is the modules functionality. However, that should, in theory, cleanup the use of objective and gain?

Comment (dxoigmn) [reply mart/attack/perturber.py:25]: As I thought about this more, it would be really nice for this to be a LitModular because then this code could be used to generate attacks using the Lightning framework.

**Comment by dxoigmn** [comment]
There is a bug introduced in b6afdd14b276987d821905a1844ec7c4df611621 with adversarial training of CIFAR10. I'm not sure where it is but this should be a good opportunity to write a test :)

File: mart/attack/perturber.py (hunk 1 above)
Comment (dxoigmn) [reply mart/attack/perturber.py]: I'll leave this for another PR.

File: mart/attack/enforcer.py [hunk 2]
```diff
@@ -84,18 +84,5 @@ def _check_constraints(self, input_adv, *, input, target):
             constraint(input_adv, input=input, target=target)
 
     @torch.no_grad()
-    def __call__(self, input_adv, *, input, target):
+    def __call__(self, input_adv, *, input, target, **kwargs):
```
Comment (dxoigmn) [inline mart/attack/enforcer.py:87]: I'm not sure this change is necessary.

File: mart/configs/attack/classification_eps1.75_fgsm.yaml [hunk 3]
```diff
@@ -20,9 +20,10 @@ optimizer:
 
 max_iters: 1
 
-perturber:
-  initializer:
-    constant: 0
+gain: "loss"
```
Comment (dxoigmn) [inline mart/configs/attack/classification_eps1.75_fgsm.yaml:23]: I think we should use the old gain functionality.

**Comment by dxoigmn** [comment]
I started breaking up this PR into smaller PRs hence the dismissed review.

File: mart/attack/adversary.py [hunk 4]
```diff
@@ -6,338 +6,116 @@
...
+            self.attacker = partial(
+                pl.Trainer,
+                num_sanity_val_steps=0,
+                logger=False,
+                max_epochs=0,
+                limit_train_batches=kwargs.pop("max_iters", 10),
+                callbacks=list(kwargs.pop("callbacks", {}).values()),  # dict to list of values
+                enable_model_summary=False,
+                enable_checkpointing=False,
+                enable_progress_bar=False,
```
Comment (mzweilin) [inline mart/attack/adversary.py:59]: I think we will fix the progress bar in a separate PR.

Comment (dxoigmn) [reply mart/attack/adversary.py:59]: Yeah it's hard to fix because you have to get the `position` correct since there are multiple nested progress bars.

**Comment by dxoigmn** [comment]
I made another batch of changes to PR to accommodate the changes in #132. As such, the tests need to be fixed. Additionally, there are some other changes related to types (e.g., relying upon `Iterable[torch.Tensor]` instead of `tuple`) that can be pulled out and made more generally useful.

File: mart/attack/callbacks/__init__.py [hunk 5]
```diff
@@ -1,4 +1,3 @@
-from .base import *
```
Comment (dxoigmn) [inline mart/attack/callbacks/__init__.py:1]: Should we just delete base.py?

File: mart/models/modular.py
Comment (dxoigmn) [inline mart/models/modular.py:1]: Pull this out into another PR.

Comment (dxoigmn) [reply mart/models/modular.py:1]: Alternatively, move `gradient_modifier` into `Perturber` via hooks.

File: mart/configs/attack/optimizer/sgd.yaml
Comment (dxoigmn) [inline mart/configs/attack/optimizer/sgd.yaml:1]: Can this be pulled into another PR?

File: mart/models/modular.py
Comment (dxoigmn) [reply mart/models/modular.py:1]: It is actually no easy to move `gradient_modifier` into `Perturber`. I think that logic should remain as close to the optimizer as possible.

File: mart/attack/adversary.py [hunk 6]
```diff
@@ -22,329 +25,171 @@
...
-            if not done:
-                on_advance_start()
-                advance()
-                on_advance_end()
-            else:
-                break
-
-        on_run_end()
-    """
+class Adversary(pl.LightningModule):
```
Comment (dxoigmn) [inline mart/attack/adversary.py:31]: Ideally this would inherit from `LitModular`. Then we can create pre-defined sequences for training.

**Comment by dxoigmn** [comment]
Why do you change the learning rate in your example?

File: mart/callbacks/progress_bar.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+__all__ = ["ProgressBar"]
+
+
+class ProgressBar(TQDMProgressBar):
+    """Display progress bar of attack iterations with the gain value."""
+
+    def init_train_tqdm(self):
+        bar = super().init_train_tqdm()
+        bar.leave = False
+        bar.set_description("Attack")
```
Comment (mzweilin) [inline mart/callbacks/progress_bar.py:19]: It looks like the progress bar is not changed.
```console
Epoch 0:  10%|██▉                | 20/200.0 [00:00<00:06, 26.93iter/s, loss=40.6]
```

File: mart/attack/adversary.py [hunk 8]
```diff
@@ -22,335 +25,177 @@
...
         self.gradient_modifier = gradient_modifier
+        self.objective_fn = objective
+        self.enforcer = enforcer
+
+        self._attacker = attacker
+
+        if self._attacker is None:
+            # Enable attack to be late bound in forward
+            self._attacker = partial(
+                pl.Trainer,
```
[the remark this answers is not in the document]
Comment (dxoigmn) [reply mart/attack/adversary.py:77]: Yeah it should inherit whatever the precision of the input is.

File: mart/callbacks/progress_bar.py (hunk 7 above)
Comment (dxoigmn) [reply mart/callbacks/progress_bar.py:19]: I never got the progress bar working for the attack.

Comment (mzweilin) [reply mart/callbacks/progress_bar.py:19]: Resolved in #153

File: mart/configs/callbacks/no_grad_mode.yaml [hunk 9]
```diff
@@ -0,0 +1,2 @@
+attack_in_eval_mode:
```
Comment (mzweilin) [inline mart/configs/callbacks/no_grad_mode.yaml:1]: Shall we make an accurate name for this?

**Review by mzweilin — APPROVED** [review]
LGTM

File: mart/configs/callbacks/no_grad_mode.yaml (hunk 9 above)
Comment (dxoigmn) [reply mart/configs/callbacks/no_grad_mode.yaml:1]: This should go away in #141.
